outline/concurrency.md

```diff
@@ -0,0 +1,84 @@
+Running tasks concurrently
+
+# Pages:
+
+* [Overview](#overview)
+* [Channels](#channels)
+* [Backpressure](#backpressure)
+
+<a name="overview"></a>
+## Overview
+
+**Status**: Unassigned.
+
+Overview of concurrency and parallelism with Tokio.
+
+#### Contents
+
+* Concurrency in an asynchronous system.
+  * Asynchronous gets you the ability to run concurrent tasks.
+  * Concurrency is not parallelism.
+    * Parallelism requires running on multiple cores.
+  * Parallelism is required to scale on modern servers
+    * Modern servers have many cores.
+* Tokio offers two runtimes, one parallel, one not.
+  * `tokio::run`
+    * Concurrent and parrallel (if the system supports it).
+    * Runs multiple OS threads.
+    * Tokio tasks are multiplexed across multiple threads.
+      * M:N scheduling.
+  * `tokio::runtime::current_thread`
+    * Concurrent, but **not** parrallel.
+    * Single OS thread.
+  * Pros / cons to both
+  * Uses Rust to enforce thread-safety.
+    * Tasks spawned on `tokio::run` must be `Send`.
+    * Tasks spawned on `runtime::current_thread` do not require `Send`.
+
+<a name="channels"></a>
+## Channels
+
+**Status**: Unassigned.
+
+Message passing for managing concurrency. This covers both oneshot and mpsc.
+
+#### Contents
+
+* Concurrency requires coordination
+  * Thread-safe as well (when parrallel).
+  * Many coordination/synchronization primitives that work in "synchronous"
+    modes of programming do not carry over to asynchronous.
+    * Mutexes don't work -- not possible to implement a future aware Mutex.
+* Message passing
+  * Primary method of coordination
+  * How it might be used in real apps.
+  * Example of message passing.
+  * Two kinds of channels, `oneshot` and `mpsc`.
```

Comment:
I assume this section will include thoughts on when to use one vs the other or will the reader be required to infer this from examples?